### PR TITLE
Fix possible memleak in PKCS7_add0_attrib_signing_time

### PR DESCRIPTION
**Pull request by bernd-edlinger**
When PKCS7_add_signed_attribute fails, the ASN1_TIME object may be leaked when it was not passed in as
input parameter.

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

File: crypto/pkcs7/pk7_attr.c [hunk 1]
```diff
@@ -98,12 +98,18 @@ int PKCS7_add_attrib_content_type(PKCS7_SIGNER_INFO *si, ASN1_OBJECT *coid)
 
 int PKCS7_add0_attrib_signing_time(PKCS7_SIGNER_INFO *si, ASN1_TIME *t)
 {
-    if (t == NULL && (t = X509_gmtime_adj(NULL, 0)) == NULL) {
+    ASN1_TIME *tmp = NULL;
```
Comment (nhorman) [inline crypto/pkcs7/pk7_attr.c:101]: Do we need the new stack variable here?  If PKCS7_add_signed_attribute fails, can't we just call ASN1_STRING_free(t) instead?

**Review by nhorman — APPROVED** [review]

**Comment by botovq** [comment]
This is an incomplete fix. Instead of just fixing an issue, it's usually worth it to fix the pattern. Make things consistent. Then you catch more of these issues. At least one more in the same file.

**Review by slontis — COMMENTED** [review]
At a minimum the documentation for X509_gmtime_adj() should be updated to indicate how you should free this created object.
Note that there is a ASN1_TIME_free() which looks like it does something else.

**Comment by bernd-edlinger** [comment]
ASN1_TIME_free would probably also work here,
but I tested the error handling with ASN1_STRING_free,
because I saw we #define ASN1_TIME ASN1_STRING

**Comment by bernd-edlinger** [comment]
The difference between ASN1_TIME_free and ASN1_STRING_free is this:
ASN1_TIME_free is rarely used.
ASN1_TIME_free does not seem to honor the str->flags ASN1_STRING_FLAG_EMBED,
but since this is not an embedded object that would not matter, and that flag
is not expected to be set if the ASN1_TIME is allocated by X509_gmtime_adj.
Shall I change here to ASN1_TIME_free?

**Comment by slontis** [comment]
If it doesnt break anything that would seem to be the matching function :)

**Comment by slontis** [comment]
Maybe see what others think before modifying it

**Comment by t8m** [comment]
I do not care either way.

**Comment by bernd-edlinger** [comment]
I would slightly prefer ASN1_TIME_free...

**Comment by t8m** [comment]
> I would slightly prefer ASN1_TIME_free...

Then change it.

**Comment by bernd-edlinger** [comment]
Done.  Thanks for reminding me...

**Review by slontis — APPROVED** [review]

**Comment by levitte** [comment]
Merged

master:
7d52539f00144cb410c4e9d8da0b9574c0badb19 Fix possible memleak in PKCS7_add0_attrib_signing_time

3.2:
1ad7f4bf979fcb5fb50013b0c9106e26c6ed2fac Fix possible memleak in PKCS7_add0_attrib_signing_time

3.1:
e83a231d3e1da1a55260503a06365b1950985933 Fix possible memleak in PKCS7_add0_attrib_signing_time

3.0:
1ef520045556c4b5ada3a6c63d32a41b5d084165 Fix possible memleak in PKCS7_add0_attrib_signing_time

**Pull request closed without merging**